physics/cires_ugwp_module_v1.F90

```diff
@@ -0,0 +1,666 @@
+!
+    dxsg =  pi2*arad/float(lonr) * knob_ugwp_ndx4lh
+    kxw  = pi2/knob_ugwp_lhmet
+!
+!   kxw  =  pi2/dxsg
+!
+! init global background dissipation for ugwp -> 4d-variable for fv3wam linked with pbl-vert_diff
+!
+
+!   allocate(fcor(latr), fcor2(latr)  )
+!
+    allocate( kvg(levs+1),   ktg(levs+1)  )
+    allocate( krad(levs+1),  kion(levs+1) )        
+    allocate( zkm(levs),   pmb(levs) )
+    allocate( rfdis(levs), rfdist(levs) )
+    
+    allocate (vert_qbo(levs))    
+    
+!
+! ak -pa  bk-dimensionless  from surf => tol_lid_pressure =0
+!
+    
+    do k=1, levs
+       pmb(k) = 1.e0*(ak(k) + pref*bk(k))    ! Pa -unit  Pref = 1.e5, pmb = Pa
+       zkm(k) = -hpskm*alog(pmb(k)/pref)
+    enddo   
+    vert_qbo(1:levs) = 0.   
+     
+   do k=kz1, kz2        		  
+      vert_qbo(k)=1.
+      if (k.le.(kz1+kz5))  vert_qbo(k)  = avqbo(k+1-kz1)
+      if (k.ge.(kz2-kz5))  vert_qbo(k) =  avqbo(kz2+1-k)
+      if (me == master) print *, 'vertqbo', vert_qbo(k), zkm(k) 		   
+    enddo
+    
+!
+! find ilaunch
+!
+    
+   do k=levs, 1, -1
+     if (pmb(k) .gt. knob_ugwp_palaunch ) exit
+   enddo
+
+      launch_level = max(k-1, 5)    ! above 5-layers from the surface
+    
+!
+! Part-1 :init_global_gwdis_v1
+!
+    call init_global_gwdis_v1(levs, zkm, pmb, kvg, ktg, krad, kion, pa_rf, tau_rf,  me,  master)
+    call rf_damp_init_v1  (levs, pa_rf, tau_rf, dtp, pmb, rfdis, rfdist, levs_rf)
+!
+! Part-2 :init_SOURCES_gws
+!
+    
+!    
+! call init-solver for "stationary" multi-wave spectra and sub-grid oro
+!
+    call init_oro_gws( knob_ugwp_wvspec(1), knob_ugwp_azdir(1), &
+         knob_ugwp_stoch(1), knob_ugwp_effac(1), lonr, kxw, cdmvgwd )
+!
+! call init-sources for "non-sationary" multi-wave spectra
+!
+    do_physb_gwsrcs=.true.
+
+    IF (do_physb_gwsrcs) THEN
+
+      if (me == master) print *, ' do_physb_gwsrcs ',  do_physb_gwsrcs, ' in cires_ugwp_init_v1 '
+      if (knob_ugwp_wvspec(4) > 0) then
+! okw
+        call init_okw_gws(knob_ugwp_wvspec(4), knob_ugwp_azdir(4), &
+                          knob_ugwp_stoch(4), knob_ugwp_effac(4), lonr, kxw )
+        if (me == master) print *, ' init_okw_gws '
+      endif
+
+      if (knob_ugwp_wvspec(3) > 0) then
+! fronts
+        call init_fjet_gws(knob_ugwp_wvspec(3), knob_ugwp_azdir(3), &
+                           knob_ugwp_stoch(3), knob_ugwp_effac(3), lonr, kxw )
+        if (me == master) print *, ' init_fjet_gws '
+      endif
+
+      if (knob_ugwp_wvspec(2) > 0) then
+! conv
+        call init_conv_gws(knob_ugwp_wvspec(2), knob_ugwp_azdir(2), &
+                           knob_ugwp_stoch(2), knob_ugwp_effac(2), lonr, kxw, cgwf )
+        if (me == master)   &
+           print *, ' init_convective GWs cgwf', knob_ugwp_wvspec(2), knob_ugwp_azdir(2)
+
+      endif
+
+     ENDIF   !IF (do_physb_gwsrcs)
+!
+!
+! Tabulated sources 
+!
+!      goto 121
+      
+      iernc=NF90_OPEN(trim(ugwp_taufile), nf90_nowrite, ncid)
+     
+       if(iernc.ne.0) then         
```

Comment:
Do not stop the model. Instead, pass the CCPP `errmsg` and `errflg` down to this routine, assign a meaningful error message to `errmsg`, set `errflg=1` and return. In the calling functions, make sure that an `errflg/=1` cascades up immediately so that CCPP can tell the host model to print the error message in its way and stop the execution.

Reply:
Done.